### PR TITLE
Progress on Upgrading Spectral Library

### PR DESCRIPTION
**Pull request by bdegley4789**
This is where I am on Issue #105 @lewismc 
I used the save() function for creating .hdr and .sli files like the AsterConversion and link you sent showed. But I'm not sure how I would query the .txt files. I may be able to do it in python as a .csv, I tried here just converting .db but that doesn't really work for query as far as I can tell

### CODE REVIEW

**Comment by bdegley4789** [comment]
This pull request doesn't need to be accepted. I just wanted to show you where I am at, so we can discuss in our meeting

File: examples/convert_to_aster.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/python
+#
+# Copyright (C) 2018 COAL Developers
```
Comment (lewismc) [inline examples/convert_to_aster.py:3]: See other copyright headers and make sure the dates are idential

File: examples/convert_to_aster.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+# Software Foundation, Inc., 51 Franklin Street, Fifth 
+# Floor, Boston, MA 02110-1301, USA.
+# encoding: utf-8
+
+'''
+Spectral Version 7 .txt file converted to ASTER .txt file format
+
+@author:     COAL Developers
+
+@copyright:  2018 COAL Developers. All rights reserved.
```
Comment (lewismc) [inline examples/convert_to_aster.py:25]: See other copyright headers and make sure the dates are idential

File: examples/convert_to_aster.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public 
+# License along with this program; if not, write to the Free 
+# Software Foundation, Inc., 51 Franklin Street, Fifth 
+# Floor, Boston, MA 02110-1301, USA.
+# encoding: utf-8
+
+'''
+Spectral Version 7 .txt file converted to ASTER .txt file format
```
Comment (lewismc) [inline examples/convert_to_aster.py:21]: ```Spectral Version 7``` doesn't refer to anything specific... make it specific.

File: pycoal/mineral.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# Copyright (C) 2017-2018 COAL Developers
```
Comment (lewismc) [inline pycoal/mineral.py:1]: Why has this been removed?

File: pycoal/mineral.py [hunk 5]
```diff
@@ -28,7 +31,7 @@ def __init__(self, library_file_name, class_names=None, threshold=0.0, in_memory
         Construct a new ``MineralClassification`` object with a spectral library
         in ENVI format such as the `USGS Digital Spectral Library 06
         <https://speclab.cr.usgs.gov/spectral.lib06/>`_ or the `ASTER Spectral
-        Library Version 2.0 <https://asterweb.jpl.nasa.gov/`_ converted with
+        Library Version 2.0 <https://speclib.jpl.nasa.gov/>`_ converted with
```
Comment (lewismc) [inline pycoal/mineral.py:34]: Why have you reverted this?

File: pycoal/mineral.py [hunk 6]
```diff
@@ -317,8 +320,8 @@ class AsterConversion:
 
     def __init__(self):
         """
-        This class provides a method for converting the `ASTER Spectral
-        Library Version 2.0 <https://asterweb.jpl.nasa.gov/>`_ into ENVI format.
+        This class provides a method for converting the ASTER Spectral
+        Library Version 2.0 into ENVI format.
```
Comment (lewismc) [inline pycoal/mineral.py:324]: Why have you reverted this?

File: examples/convert_to_aster.py (hunk 3 above)
Comment (lewismc) [reply examples/convert_to_aster.py:21]: Additionally, **THERE IS NO ASTER .txt FILE FORMAT**... the input is a DB file. I've already explained this several times.

File: examples/convert_to_aster.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+@author:     COAL Developers
+
+@copyright:  2018 COAL Developers. All rights reserved.
+
+@license:    GNU General Public License version 2
+
+@contact:    coal-capstone@googlegroups.com
+'''
+
+from argparse import ArgumentParser
```
Comment (lewismc) [inline examples/convert_to_aster.py:32]: You are not using this import... remove it

File: examples/convert_to_aster.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+from argparse import RawDescriptionHelpFormatter
```
Comment (lewismc) [inline examples/convert_to_aster.py:33]: You are not using this import... remove it

File: examples/convert_to_aster.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import sys
```
Comment (lewismc) [inline examples/convert_to_aster.py:34]: Remove any unused imports... this is basic stuff.

File: examples/convert_to_aster.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+import os
+from sys import path
+from os import getcwd
+import inspect
+import math
+import numpy
+
+#This will convert a Spectral Version .txt file to a ASTER .txt file
+library_filename = 'splib07a_Alizarin_crimson_(dk)_GDS780_ASDFRa_AREF.txt'
```
Comment (lewismc) [inline examples/convert_to_aster.py:43]: Where does this file come from?

File: examples/convert_to_aster.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+with open(library_filename,'r') as f:
+    for i, l in enumerate(f):
```
Comment (lewismc) [inline examples/convert_to_aster.py:45]: Why are you creating the variables ```i``` and ```l``` here?

File: examples/convert_to_aster.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        pass
+
+f = open(library_filename,'r')
+nameTemp = f.readline()
```
Comment (lewismc) [inline examples/convert_to_aster.py:49]: Variable ```nameTemp``` is a terrible name for this variable. Please see the [basics of variable naming in Python](https://en.wikipedia.org/wiki/Naming_convention_(programming)#Python_and_Ruby)

File: examples/convert_to_aster.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+name = nameTemp[23:]
+k = 0
+x = open('2.txt','w')
```
Comment (lewismc) [inline examples/convert_to_aster.py:52]: ```2.txt``` is a ridiculous name for any file however seeing as it is a _temporary_ file AND you clean it up I suppose this is OK.

File: examples/convert_to_aster.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+while(k < i):
+    num = float(f.readline()) * 100
+    num = num / 1000
+    num2 = num * 10
+    num = round(num,4)
```
Comment (lewismc) [inline examples/convert_to_aster.py:57]: What is your decision to round these floating point numbers?

File: examples/convert_to_aster.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    x.write(line)
+    x.write('\n')
+    k = k+1
+f = open(library_filename,'w')
+f.write('Name:')
+f.write(name)
+f.write('Type:\n')
+f.write('Class:\n')
+f.write('Subclass:\n')
+f.write('Particle Size:  Solid\n')
```
Comment (lewismc) [inline examples/convert_to_aster.py:69]: This all looks like garbage to me. I don't understand why you are deciding to write static final string values for certain fields here.

File: examples/convert_to_aster.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/python
```
Comment (lewismc) [inline examples/convert_to_aster.py:1]: This is not even a generic file for it to be named ```convert_to_aster.py```... look at the code you've provided. It is intrinsically tied to a very specific spectra... how can this be generic?

File: examples/convert_to_aster.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+while(j < i):
+    num = x.readline()
+    f.write(num)
+    j = j+1
+f.close()
+x.close()
+os.rename(library_filename,library_filename + '.spectrum.txt')
+os.remove('2.txt')
+print("Successfully converted file " + library_filename)
+
```
Comment (lewismc) [inline examples/convert_to_aster.py:103]: Why have you added two lines on the bottom?

File: examples/convert_to_aster.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+j = 0
+x.close()
+x = open('2.txt','r')
+while(j < i):
+    num = x.readline()
+    f.write(num)
+    j = j+1
+f.close()
+x.close()
+os.rename(library_filename,library_filename + '.spectrum.txt')
```
Comment (lewismc) [inline examples/convert_to_aster.py:100]: Why is this filename now called ```blah.spectrum.txt```? It has nothing to do with a spectrum

File: examples/example_splib07.py [hunk 19]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/python
+#
+# Copyright (C) 2018 COAL Developers
```
Comment (lewismc) [inline examples/example_splib07.py:3]: Look at every other file header.... make sure they match identically.

File: examples/example_splib07.py [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public 
+# License along with this program; if not, write to the Free 
+# Software Foundation, Inc., 51 Franklin Street, Fifth 
+# Floor, Boston, MA 02110-1301, USA.
+# encoding: utf-8
+
+'''
+example_sam -- Plot Spectral Angles
```
Comment (lewismc) [inline examples/example_splib07.py:21]: This is NOT ```example_sam``` and you are NOT plotting spectral angles.

File: examples/example_splib07.py [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+
+.sli and .hdr files will be generated from spectral library.
+
+@author:     COAL Developers
+
+@copyright:  2018 COAL Developers. All rights reserved.
```
Comment (lewismc) [inline examples/example_splib07.py:27]: Make sure file headers are identical.

File: examples/example_splib07.py [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+import pycoal
+sys.path.insert(0, '../pycoal')
+reload(sys)
+sys.setdefaultencoding('utf8')
+import mineral
+import math
+import numpy
+import spectral
+
+# load library
```
Comment (lewismc) [inline examples/example_splib07.py:54]: Why do you have all of this commented out code? Either use it in the implementation or remove it.

File: examples/splib07_convert.py [hunk 23]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/python
+#
+# Copyright (C) 2018 COAL Developers
```
Comment (lewismc) [inline examples/splib07_convert.py:3]: Make sure license headers are identical.

File: examples/splib07_convert.py [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public 
+# License along with this program; if not, write to the Free 
+# Software Foundation, Inc., 51 Franklin Street, Fifth 
+# Floor, Boston, MA 02110-1301, USA.
+# encoding: utf-8
+
+'''
+Spectral Version 7 .txt file converted to ASTER .txt file format
```
Comment (lewismc) [inline examples/splib07_convert.py:21]: See previous comment

File: examples/splib07_convert.py [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+
+@author:     COAL Developers
+
+@copyright:  2018 COAL Developers. All rights reserved.
```
Comment (lewismc) [inline examples/splib07_convert.py:25]: See previous comment

File: examples/splib07_convert.py [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+
+@license:    GNU General Public License version 2
+
+@contact:    coal-capstone@googlegroups.com
+'''
+
+from argparse import ArgumentParser
+from argparse import RawDescriptionHelpFormatter
+
+import logging
```
Comment (lewismc) [inline examples/splib07_convert.py:35]: Remove ALL unused logging.

File: pycoal/mineral.py [hunk 27]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
         library = aster_database.create_envi_spectral_library(spectrum_ids, band_info)
 
         library.save(hdr_file)
+
+class SpectralToAsterConversion:
+    
+    def __init__(self):
+        """
+            This class provides a method for converting Spectral .txt files into
+            ASTER Library Version 2.0 .txt files
```
Comment (lewismc) [inline pycoal/mineral.py:374]: Hyperlink ASTER

File: pycoal/mineral.py [hunk 28]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
         library = aster_database.create_envi_spectral_library(spectrum_ids, band_info)
 
         library.save(hdr_file)
+
+class SpectralToAsterConversion:
+    
+    def __init__(self):
+        """
+            This class provides a method for converting Spectral .txt files into
```
Comment (lewismc) [inline pycoal/mineral.py:373]: Hyperlink USGS Spectral Library Version 7

File: pycoal/mineral.py [hunk 29]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
...
+            
+            Args:
+                none
+            """
+        pass
+    
+    @classmethod
+    def convert(cls, library_filename=""):
+        """
+            This class method converts a Spectral version 7 .txt file into
```
Comment (lewismc) [inline pycoal/mineral.py:384]: Hyperlink USGS Spectral Library Version 7

File: pycoal/mineral.py [hunk 30]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
...
+            an ASTER Library Version 2.0 .spectrum.txt file
```
Comment (lewismc) [inline pycoal/mineral.py:385]: This comment is useless as no-ne knows what a spectrum.txt file is

File: pycoal/mineral.py [hunk 31]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
...
+            
+            Spectral Library Version 7 can be downloaded here https://speclab.cr.usgs.gov/spectral-lib.html
```
Comment (lewismc) [inline pycoal/mineral.py:387]: Use hyperlink markup.

File: pycoal/mineral.py [hunk 32]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
...
+            
+            How to use on entire USGS Spectral Library Version 7:
+            1. Move all .txt files in Spectral Library Version 7 to one directory named 'usgs_splib07_modified'
```
Comment (lewismc) [inline pycoal/mineral.py:390]: This logic is flawed... the user will be sitting waiting doing this forever. The Python code should take care of doing this. It is called automation.

File: pycoal/mineral.py [hunk 33]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
...
+            2. Place this directory('usgs_splib07_modified') you just made in the pycoal/examples directory
```
Comment (lewismc) [inline pycoal/mineral.py:391]: Why in Gods name would we ever want to do that.
**WE ARE TRYING TO USE THIS TOOLKIT FROM THE COMMAND LINE**... come on.

File: pycoal/mineral.py [hunk 34]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
...
+            3. run splib07_convert.py in the pycoal/examples directory
```
Comment (lewismc) [inline pycoal/mineral.py:392]: No no no

File: pycoal/mineral.py [hunk 35]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
...
+            4. run example_splib07.py in the pycoal/examples directory
```
Comment (lewismc) [inline pycoal/mineral.py:393]: No no no this is riduculous

File: pycoal/mineral.py [hunk 36]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
...
+            5. s07av95a_envi.hdr and s07av95a_envi.sli will be generated
```
Comment (lewismc) [inline pycoal/mineral.py:394]: Where will they be generated? Which directory? You NEED TO COMMENT YOUR CODE.

File: pycoal/mineral.py [hunk 37]
```diff
@@ -362,3 +365,96 @@ def convert(cls, data_dir="", db_file="", hdr_file=""):
...
+        f.write('Name:')
+        f.write(name)
+        f.write('Type:\n')
+        f.write('Class:\n')
+        f.write('Subclass:\n')
+        f.write('Particle Size:  Solid\n')
+        f.write('Sample No.:  0095UUUASP\n')
+        f.write('Owner:\n')
+        f.write('Wavelength Range:  ALL\n')
+        f.write('Origin: Spectra obtained from the Noncoventional Exploitation Factors\n')
```
Comment (lewismc) [inline pycoal/mineral.py:431]: This is useless and only works for one specific file. This needs to change entirely.

**Review by lewismc — CHANGES_REQUESTED** [review]
@bdegley4789 see my comments in their entirety. Once you've seen them, close this pull request and start again. Next time around, please follow the logic that I give you when we have our meetings.

**Comment by bdegley4789** [comment]
Hey @lewismc 

I will definitely update the headers, add comments to my code and remove any commented out code like you said.

I also agree it is a bad method to have users manually move files

I can use os.walk to loop through directories so users can do everything from the command line. 
I tried doing this but couldn't figure it out so I just moved all the .txt files to one directory and used os.listdir() for the time being.
I will work on incorporating these changes

**Comment by bdegley4789** [comment]
There is confusion on the rest of my code.

The AsterConversion class in mineral.py has two ways of working.
One way is what you talk about with input as a DB file which you mentioned in your comment
The second way is by getting a directory of .spectrum.txt files and generating a .db file

Look at AsterConversion class in mineral.py on Line 339 mineral.py 
https://github.com/capstone-coal/pycoal/blob/41bed1807a2d1d798c5bc8ee79a4b5cc336b7b8a/pycoal/mineral.py#L339

https://github.com/capstone-coal/pycoal/blob/41bed1807a2d1d798c5bc8ee79a4b5cc336b7b8a/pycoal/mineral.py#L340

https://github.com/capstone-coal/pycoal/blob/41bed1807a2d1d798c5bc8ee79a4b5cc336b7b8a/pycoal/mineral.py#L341

For this method I used the Second way described above ^

What I did is
1) Take the Aster Library which is all .spectrum.txt files
2) Then generate a .db file from the .spectrum.txt files
3) Then generate a .sli and .hdr file from the .db file

I changed change the Spectral V7 .txt files to match the Aster .spectrum.txt file format, while still using it's original name and values.
Then I used the method of 2 & 3 above to generate the .sli and .hdr files for Spectral Library Version 7

Remember, our current AsterConverion only looks at the name and the query for the x column which is the wavelength values which I put in that same column.

Line 355 mineral.py
https://github.com/capstone-coal/pycoal/blob/41bed1807a2d1d798c5bc8ee79a4b5cc336b7b8a/pycoal/mineral.py#L355 

The other values that I put in are never used for .sli and .hdr creation in regards to your last inline comment. So, this method will work for all files

**Comment by bdegley4789** [comment]
convert_to_aster.py was used for testing the convert function and you are correct it only works on one file, that file is just a random file from the spectral library I used to work with. 
That is why I didn't include convert_to_aster.py in those 5 steps I wrote.

splib07_convert.py will work on the entire directory you provide

mineral.py line 402
https://github.com/bdegley4789/pycoal/blob/7b10aa9592ea913a4816cdab49aa2a3754b3a64f/pycoal/mineral.py#L402

https://github.com/bdegley4789/pycoal/blob/7b10aa9592ea913a4816cdab49aa2a3754b3a64f/pycoal/mineral.py#L403

https://github.com/bdegley4789/pycoal/blob/7b10aa9592ea913a4816cdab49aa2a3754b3a64f/pycoal/mineral.py#L404


That for loop and variables i and l are used to find the number of lines in the file so I know how much data I need to record and store

The code does work on every .txt file in the Spectral Library which I tested. 
If you look at the hdr files for splib06 and splib07 I staged in https://drive.google.com/drive/folders/1YVhdLxvrZE3eC97OEXathLMJRgWt8haO they are the same format. 
I built this method around using the .save function you had sent me and looking at the Aster conversion code provided and refrenced in the issue discussed in the meetings.
I might need to tweek the name of the mineral values and edit the code in mining, mineral and environment.py to look for the new names of the spectral library 7 minerals
Other than that it looks like it is all working correctly with the method I have used. I think we should go forward with this after I make the changes mentioned in my prior prior comment

**Comment by bdegley4789** [comment]
I'll wait to close this pull request, if you still want to, after reviewing my comments

**Comment by lewismc** [comment]
@bdegley4789 thanks for considering the comments and criticisms.

> I will definitely update the headers, add comments to my code and remove any commented out code like you said.

Yes please go ahead and do this before we move forward. Basically lets start from there.

After that, then address the following

> I can use os.walk to loop through directories so users can do everything from the command line.
I tried doing this but couldn't figure it out so I just moved all the .txt files to one directory and used os.listdir() for the time being. I will work on incorporating these changes

**Pull request closed without merging**